crates/lib/src/qpu/api.rs

```diff
@@ -91,74 +72,287 @@ impl From<String> for JobId {
+///       overrides the `quantum_processor_id` parameter.
 pub async fn retrieve_results(
     job_id: JobId,
-    job_target: &JobTarget,
+    quantum_processor_id: String,
     client: &Qcs,
-) -> Result<ControllerJobExecutionResult, GrpcClientError> {
+    execution_options: &ExecutionOptions,
+) -> Result<ControllerJobExecutionResult, QpuApiError> {
     #[cfg(feature = "tracing")]
-    tracing::debug!("retrieving job results for {} on {}", job_id, job_target,);
+    tracing::debug!(
+        "retrieving job results for {} on {} using options {:?}",
+        job_id,
+        quantum_processor_id,
+        execution_options,
+    );
 
     let request = GetControllerJobResultsRequest {
         job_execution_id: job_id.0,
-        target: Some(job_target.into()),
+        target: Some(
+            execution_options
+                .connection_strategy
+                .get_results_target(&quantum_processor_id),
+        ),
     };
 
-    let mut controller_client = match job_target {
-        JobTarget::EndpointId(endpoint_id) => {
-            client
-                .get_controller_client_with_endpoint_id(endpoint_id)
-                .await
-        }
-        JobTarget::QuantumProcessorId(quantum_processor_id) => {
-            client.get_controller_client(quantum_processor_id).await
-        }
-    }?;
+    let mut controller_client = execution_options
+        .connection_strategy
+        .get_controller_client(client, &quantum_processor_id)
+        .await?;
 
-    controller_client
+    Ok(controller_client
         .get_controller_job_results(request)
-        .await?
+        .await
+        .map_err(GrpcClientError::RequestFailed)?
         .into_inner()
         .result
-        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))
+        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))?)
+}
+
+/// Options avaialable when executing a job on a QPU.
+///
+/// Use [`Default`] to get a reasonable set of defaults, or start with [`ExecutionOptionsBuilder`]
+/// to build a custom set of options.
+#[derive(Builder, Clone, Debug, Default, PartialEq, Eq)]
+pub struct ExecutionOptions {
+    #[doc = "The [`ConnectionStrategy`] to use to establish a connection to the QPU."]
+    connection_strategy: ConnectionStrategy,
+}
+
+impl ExecutionOptions {
+    /// Get the [`ConnectionStrategy`]
+    #[must_use]
+    pub fn connection_strategy(&self) -> &ConnectionStrategy {
+        &self.connection_strategy
+    }
+}
+
+/// The connection strategy to use when submitting and retrieving jobs from a QPU.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub enum ConnectionStrategy {
+    /// Connect through the publicly accessible gateway.
+    #[default]
+    GatewayOnly,
+    /// Connect directly to the default endpoint, bypassing the gateway. Should only be used when you
+    /// have direct network access and an active reservation.
+    DirectAccess,
```

Comment:
(this would be a better place for the note about name consistency, `Only` vs ``

Noting that the docstrings are slightly different here from the python bindings ("always"). Those seem a little more specific.